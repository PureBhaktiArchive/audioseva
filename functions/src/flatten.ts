/*!
 * sri sri guru gauranga jayatah
 */

/**
 * Flattens nested objects so that the resulting object is one-level.
 * Property names are combined with forward slash as separator.
 * This function is useful for preparing an object for Firebase RTDB `update` function.
 *
 * @param source object to flatten
 * @param prefix @string prefix to add to the property path
 * @param target target object
 */
export const flatten = (source, prefix = '', target = {}) =>
  Object.entries(source).reduce((memo, [key, val]) => {
    const nestedKey = `${prefix}${key}`;
    if (typeof val === 'object' && val && val['.sv'] === undefined) {
      flatten(val, `${nestedKey}/`, memo);
    } else {
      target[nestedKey] = val;
    }
    return memo;
  }, target);
