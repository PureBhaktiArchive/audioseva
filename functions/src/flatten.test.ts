/*!
 * sri sri guru gauranga jayatah
 */

import { flatten } from './flatten';

describe('flatten', () => {
  test('should flatten pure object', () => {
    const source = {
      root1: {
        second1: {
          leaf1: 'string',
          leaf2: 108,
          timestamp: {
            '.sv': 'TIMESTAMP',
          },
        },
      },
      root2: {
        second2: ['value1', 'value2', 3],
      },
      root3: 'value right away',
      root4: null,
      root5: undefined,
      root6: 0,
      root7: '',
      root8: NaN,
    };

    expect(flatten(source, 'prefix/')).toMatchSnapshot();
  });
});
