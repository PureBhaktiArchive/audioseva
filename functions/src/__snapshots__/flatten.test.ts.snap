// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flatten should flatten pure object 1`] = `
Object {
  "prefix/root1/second1/leaf1": "string",
  "prefix/root1/second1/leaf2": 108,
  "prefix/root1/second1/timestamp": Object {
    ".sv": "TIMESTAMP",
  },
  "prefix/root2/second2": Array [
    "value1",
    "value2",
    3,
  ],
  "prefix/root3": "value right away",
  "prefix/root4": null,
  "prefix/root5": undefined,
  "prefix/root6": 0,
  "prefix/root7": "",
  "prefix/root8": NaN,
}
`;
